### ISSUE TITLE
Mention Spaces when not start of text not working

### ISSUE THREAD

**Issue by jameshays** [nathantannar4/InputBarAccessoryView#48]
I've spent the day in the code trying to figure out a solution for this and haven't, so I'll throw it back to the author for guidance and hopefully a solution.

In my app and the Example app:
When I type `@Rick S` I get autocompletion results as expected.

When I type `Hey @Rick` I get autocompletion results as expected

When I add a space and type `Hey @Rick ` or `Hey @Rick S`, I don't get results.  

A couple of things I'm seeing.  
1) I think this code `session.spaceCounter += spaceCountDiff` on lie 410 of AutocompleteManager should read `session.spaceCounter = spaceCountDiff`.  Otherwise, I don't see where session.spaceCounter ever decreases or resets to 0.  It always grows.  If it always grows, this line `if let session = currentSession, session.spaceCounter <= maxSpaceCountDuringCompletion` on 164 in AutocompleteManager always fails and the session is unregistered

2) If I set `session.spaceCount = spaceCountDiff`, the code is then allowed to pass that mentioned if statement and the .whitespaces character set is removed from the copy of delimiters.  This then calls textField.find with the set of prefixes and the reduced delimiter set.  Since it no longer breaks on white spaces, the wordAtCarat function returns the entire string, causing `result.word.hasPrefix(prefix)` on line 20 of UITextView+Extensions to fail.

Any thoughts on how to resolve this?

**Comment by jameshays** [comment]
This is my solution at the moment.

```internal extension UITextView {
    
    func find(prefixes: Set<String>, with delimiterSet: CharacterSet) -> (prefix: String, word: String, range: NSRange)? {
        guard prefixes.count > 0
            else { return nil }
        
        return prefixes.compactMap({
            guard let prefix = $0.first else { return nil }
            return find(prefix: prefix, with: delimiterSet)
        }).last
    }
    
    func find(prefix: Character, with delimiterSet: CharacterSet) -> (prefix: String, word: String, range: NSRange)? {
        guard let caretRange = self.caretRange,
            let cursorRange = Range(caretRange, in: text) else { return nil }
        
        var substring = text[..<cursorRange.upperBound]
        guard let prefixIndex = substring.lastIndex(of: prefix) else { return nil }
        
        let wordRange: Range = prefixIndex..<cursorRange.upperBound
        substring = substring[wordRange]
        
        let location = wordRange.lowerBound.encodedOffset
        let length = wordRange.upperBound.encodedOffset - location
        let range = NSRange(location: location, length: length)
        
        return (String(prefix), String(substring), range)
    }
    
    func wordAtCaret(with delimiterSet: CharacterSet) -> (word: String, range: NSRange)? {
        guard let caretRange = self.caretRange,
            let result = text.word(at: caretRange, with: delimiterSet)
            else { return nil }
        
        let location = result.range.lowerBound.encodedOffset
        let range = NSRange(location: location, length: result.range.upperBound.encodedOffset - location)
        
        return (result.word, range)
    }
    
    var caretRange: NSRange? {
        guard let selectedRange = self.selectedTextRange else { return nil }
        return NSRange(
            location: offset(from: beginningOfDocument, to: selectedRange.start),
            length: offset(from: selectedRange.start, to: selectedRange.end)
        )
    }
    
}
```

After more testing, I may create a PR for it.

**Comment by nathantannar4** [comment]
@jameshays Thanks for helping out with this! Tbh I was in a crunch when I needed this for a project of mine and will be the first to admit it was not robust enough for all cases, but it worked for what I needed

I would accept PRs

**Comment by jameshays** [comment]
In the autocomleteManager, I have this as well.

```
    public func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        
        // Ensure that the text to be inserted is not using previous attributes
        preserveTypingAttributes()
        
        if let session = currentSession {
            let textToReplace = (textView.text as NSString).substring(with: range)
            let deleteSpaceCount = textToReplace.filter { $0 == .space }.count
            let insertSpaceCount = text.filter { $0 == .space }.count
            let spaceCountDiff = insertSpaceCount - deleteSpaceCount
            session.spaceCounter = spaceCountDiff
        }
...
```

session.spaceCount = spaceCountDiff.  it doesn't seem to always be accurate when deleting spaces and I haven't worked through that yet.  It seems to be working for now though.  More testing is probably warranted.

**Comment by jameshays** [comment]
I created a PR here.  https://github.com/nathantannar4/InputBarAccessoryView/pull/50

Give it a go and let me know what you find.

**Comment by nathantannar4** [comment]
Fixed in 4.2.1
